src/ert/ensemble_evaluator/evaluator_async.py

```diff
@@ -0,0 +1,476 @@
+import asyncio
+import logging
+import pickle
+from collections import OrderedDict
+from contextlib import asynccontextmanager, contextmanager
+from http import HTTPStatus
+from typing import (
+    Any,
+    AsyncIterator,
+    Awaitable,
+    Callable,
+    Dict,
+    Generator,
+    Iterable,
+    List,
+    Optional,
+    Set,
+    Tuple,
+)
+
+import cloudevents.exceptions
+import cloudpickle
+import websockets
+from cloudevents.conversion import to_json
+from cloudevents.http import CloudEvent, from_json
+from websockets.datastructures import Headers, HeadersLike
+from websockets.exceptions import ConnectionClosedError
+from websockets.legacy.server import WebSocketServerProtocol
+
+from ert.serialization import evaluator_marshaller, evaluator_unmarshaller
+
+from ._builder import Ensemble
+from .config import EvaluatorServerConfig
+from .identifiers import (
+    EVGROUP_FM_ALL,
+    EVTYPE_EE_SNAPSHOT,
+    EVTYPE_EE_SNAPSHOT_UPDATE,
+    EVTYPE_EE_TERMINATED,
+    EVTYPE_EE_USER_CANCEL,
+    EVTYPE_EE_USER_DONE,
+    EVTYPE_ENSEMBLE_CANCELLED,
+    EVTYPE_ENSEMBLE_FAILED,
+    EVTYPE_ENSEMBLE_STARTED,
+    EVTYPE_ENSEMBLE_STOPPED,
+)
+from .snapshot import PartialSnapshot
+from .state import (
+    ENSEMBLE_STATE_CANCELLED,
+    ENSEMBLE_STATE_FAILED,
+    ENSEMBLE_STATE_STOPPED,
+)
+
+logger = logging.getLogger(__name__)
+
+
+class EnsembleEvaluatorAsync:
+    def __init__(self, ensemble: Ensemble, config: EvaluatorServerConfig, iter_: int):
+        # Without information on the iteration, the events emitted from the
+        # evaluator are ambiguous. In the future, an experiment authority* will
+        # "own" the evaluators and can add iteration information to events they
+        # emit. In the meantime, it is added here.
+        # * https://github.com/equinor/ert/issues/1250
+        self._iter: int = iter_
+        self._config: EvaluatorServerConfig = config
+        self._ensemble: Ensemble = ensemble
+
+        self._loop: Optional[asyncio.AbstractEventLoop] = None
+        self._done: asyncio.Future[bool] = asyncio.Future()
+
+        self._clients: Set[WebSocketServerProtocol] = set()
+        self._dispatchers_connected: asyncio.Queue[None] = asyncio.Queue()
+
+        self._events: asyncio.Queue[CloudEvent] = asyncio.Queue()
+        self._messages_to_send: asyncio.Queue[str] = asyncio.Queue()
+
+        self._result = None
+
+        self._ee_tasks: List[asyncio.Task[None]] = []
+        self._server_started: asyncio.Event = asyncio.Event()
+
+        # batching section
+        self._batch_processing_queue: asyncio.Queue[
+            List[Tuple[Callable[[List[CloudEvent]], Awaitable[None]], CloudEvent]]
+        ] = asyncio.Queue()
+        self._max_batch_size: int = 500
+        self._batching_interval: int = 2
+
+    async def _publisher(self) -> None:
+        try:
+            while True:
+                msg = await self._messages_to_send.get()
```

Comment:
Can we change `msg` to `msg_to_send`? It is a bit confusing because we use `...get()` on `messages_to_send`

Reply:
Can we change msg to msg_to_send here too?

Reply:
I think this is fine in this case, we we are polling the msg from the queue.